Migrate from Vue 2 to Vue 3

gerteck opened the pull request:
**What is the purpose of this pull request?**

- [x] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [x] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- Others, please explain:

### **Overview of PR:**

Fixes #2597 - Steps Needed for Vue Migration. 

MarkBind (v5.5.3) is currently using Vue 2. However, Vue 2 has reached EOL in 2023 and limits the extensibility and maintainability of MarkBind, especially the vue-components package, and external dependencies for Vue 2 components that are no longer actively maintained. (UI Library Package). This PR migrates MarkBind from Vue 2 to Vue 3. This builds on extensive efforts from #2224 - Vue Migration attempt 2 years ago by @EltonGohJH, which has helped immensely to set the foundation for this PR. 

There are no functional changes to how MarkBind behaves:
* SSR is preserved and (fully) functional
* Hydration is preserved and (fully) functional
* Vue components preserved using Options API (vs. Composition API)

### **Overview of changes:**

Some basic understanding of both Vue 2, Vue 3 and a good understanding of MarkBind (in particular with regards to SSR, Hydrating, Bundling) is needed to review? the PR. I have previously linked some hopefully helpful resources in issue #2597 to help in understanding this PR.

The Vue migration involves major changes in 3 out of 4 packages of Markbind, namely the `core`, `core-web`, and `vue-components`. Changes in `cli` are only functional snapshot tests. Changes have been separated into commits that are aim to be modular and digestable.

**Commit History Overview**
1. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/9252c201840c422659c047e88586f00fa088781e): `Adapt processing logic to use Vue 3.3.11` - vue is updated for the core library. 
    * This entails changes in API to both the [renderFunction](https://www.npmjs.com/package/@vue/compiler-sfc) generation function, and the [renderToString](https://www.npmjs.com/package/@vue/server-renderer) function. There is no documentation for the renderFunction generation function, you need to directly reference the type definitions from [vuejs source code](https://github.com/vuejs/core/blob/main/packages/compiler-core/src/options.ts) as specified by npm package description. 
    * I also update the bundled vue binary to V3.3.11 here.

2. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/2ec3b288ad0e0649e2fcb13b2775b2d6567f6897): `Migrate Vue Components` - I update the configurations for Vue Components to Vue3 (test, lint). 
    * I also update configuration dependencies in the `vue-components` package (`vue-style-loader`, `jest-env-jsdom`). 
    * Next, I update minimal syntax changes in all components (and directives). Reviewers can refer to the [vue migration guide](https://v3-migration.vuejs.org/) to look at syntax changes or API changes that are mentioned there. 
    * Naturally, vue component dependencies are also updated (`floating-vue`, `vue-final-modal`, `portal-vue`). Every component or configuration dependency updated probably has some separate form of migration guide online. 
    * Changes are largely minimal.

3. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/6bcf6dbd71e2386267458283a56509416cb6680f): `Update core-web package 1` - Update dependencies for core-web (bundling). Update configuration for webpack. 

4. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/450d78b35627604b3d4eb60652bd4956a52c97f4): `Update core-web package 2` - Update the essential webpage js logic `index.js` in core-web. 
    * Simply put, this entails using the new renderFunction that we previously generated in commit 1 and registering the vue components in commit 2. 
    * Additionally, I fixed a bug in the webpack config that I did not detect previously. (see commit message for details). 
    * Binaries for Vue components and markbind.js are regenerated.

5. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/0486230ee73c210f5b830c46c3634c3b9d64ab15): `Fix Hydration Issue for Question component` - Fixed issue that I did not detect previously, related to commit 2.

6. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/a7af11a5e604dc422c21326b6d21251325ddfc56): `Update Vue Testcases and Snapshots` - As per title. I also realized ( #2610 ), so I amended the erroneous tests. No point keeping useless tests. 
    * There are some API changes to the testing so they were updated.

7. [Commit](https://github.com/MarkBind/markbind/commit/0b93db3b0f82288d41463729243391eb4bc75c27): `Merge master`: PR updated with necessary changes from pre-requisite PRs. 
    * Basically, the PRs decouples plugins from using Vue. (They were using a workaround that works with Vue 2, but can't be done with Vue 3.) Thanks to these PRs, we don't need to worry about vue migration affecting plugins. Coincidentally, the PRs also enhance efficiency and fix other bugs identified. 

8. Update license txt file as per [MIT License](https://opensource.org/license/mit)

9. [Commit](https://github.com/MarkBind/markbind/pull/2622/commits/7ea3d6d544e7b0d3b69333f2d88279d793d67da5): `Update CLI Snapshots` - Snapshots for test generated sites are regenerated and commited. 
    * Don't be intimidated - actually there's no real change in most if not all of the html files. It's only because I updated the vue.js file name to better reflect the vue version. (which is awesome - means no change in HTML rendered!)
    *  Ignore the changes to the renderfunction binary (`*.renderjs`), it is not meaningful. I realized that the render function is unnecessarily compared. This causes a lot of issues like pollution, unnecessary increased test load. I will see if I can make a separate PR to address this issue to increase developers quality of life. Related to #1637.

10. Congratulations for getting this far. There's no other unrelated initial changes. (Further changes and bug fixes are detailed in comment history below, or resolved comments in code review)

Changelog in comment history:
> **Changelog** (Beyond initial PR changes):
> * Built the CS2103 website on gh-pages (serves updated version)
> * Update modal opening animation (update breaking syntax)
> * Update dropdown functionality (hide submenus - update breaking syntax with new logic)
> * Fix tree diagram functionality (whitespace issue - vue breaking change)
> * Fix spacing between elements (whitespace issue - vue breaking change)
> * Investigate render warning (invalid HTML in docs)
> * Fix to images not loaded, panel somehow missing (register plugins in new Vue instance, Retriever.vue)
> * Updated fix to navbar content not properly handled (update breaking vue CSS scoped syntax)
> * Update Question hydration error - (disable animations for this PR for now)
> * Add docs for how MarkBind uses Vue & Update outdated docs on vue 
> * Fix whitespace issue in site-nav / page-nav (update portal-vue breaking syntax)
> * Closeable Directive fixed (update hook)
> * Fixed Pic component on load computation of height and width
> * Fixed Thumbnail and Annotate components `@load` handling case where it does not fire.

---

### Issues Fixed

Fixes #2597 - Steps Needed for Vue Migration
Fixes #2084 - Investigate Vue Migration

Fixes #2610 - removed irrelevant ineffective tests that don't work.

Fixes #2171 - reloading logic abstracted into `savePageRenderFnForHotReload` method in `PageVueServerRenderer`, to reduce coupling from `Page/index.js`.

Fixes #1536 - scopedSlots have been replaced to slots as per Vue 3 migration

---

### Others

**Related PRs for Reference**:
https://github.com/MarkBind/markbind/pull/2596 - Unformatted Vue Migration PR (Closed)
https://github.com/MarkBind/markbind/pull/1534 - Vue 2 SSR to Resolve FOUC

**Related PRs (not needed for Reference)**:
Additionally, I have tried as best as possible to isolate any separable changes into separate PRs from this PR for ease of code review and understanding. It is not necessary to look through those changes. These PRs, which include both enhancement and code maintanence, include:
* https://github.com/MarkBind/markbind/pull/2625 - Refactor codeBlockButtonPlugins
* https://github.com/MarkBind/markbind/pull/2626 - Update dataTable
* https://github.com/MarkBind/markbind/pull/2614 - Update Mermaid

I don't think it is feasible to further break down this PR (tests would not work).

Additionally, during the migration process, I have identified some issues, such as #2610, #2612, #2613, #2615, #2623, #2627, #2629, #2630, #2631, #2624. Most of these issues are not addressed in the scope of this PR, but a review of this PR might provide context to tackle these issues separately. Also, there are definitely regressions or changes due to change in dependencies, especially for the vue components. These can also be tackled separately in the future. Appreciate the help to identify any regressions if spotted and raising appropriate issues after the Vue Migration has been merged. 

To-do:
* Update documentation on Vue related sections. From what I can see, the writing components section. However, actually I think the section is general enough that it doesn't really need much changes for correctness. Maybe I will raise a separate issue and add on to the documentation and make changes separately. idk. thoughts?

---

**Testing instructions:**

Some general tests I could think of:

1. **Run Tests**
   - `npm run test` for `core`, `cli`, and `vue-components`.

2. **Build & Serve Sample Site**
   - `markbind init`, `markbind build`, `markbind serve`.
   - Verify components render correctly and interact as expected.

3. **Check SSR & Hydration**
   - Confirm no hydration mismatch or console warnings.
   - Verify correct server-rendered HTML and client hydration.

4. **Plugin Compatibility**
   - Test various plugins to ensure migration has no side effects.

5. **Cross-browser Check**
   - Open site in Chrome, Firefox, Safari, and Edge.

---

**Proposed commit message: (wrap lines at 72 characters)**

Migrate from Vue 2 to Vue 3

Vue 2 has reached end-of-life in 2023, limiting maintainability,
extensibility of MarkBind. This commit migrates the codebase 
(core, CLI, and vue-components) to Vue 3.3.11.

The migration includes major updates to server-side rendering
logic, hydration, component syntax, bundling setup (Webpack), 
and test  configurations. Vue 2 dependencies and deprecated 
APIs have been replaced accordingly. Snapshot tests and test 
site output have also been updated to reflect new Vue version.

This migration improves future extensibility and compatibility
with modern tooling and Vue ecosystem packages.

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes (Best effort) <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

---

**Reviewer checklist:**

Indicate the [SEMVER](https://semver.org/) impact of the PR:
- [ ] Major (when you make incompatible API changes)
- [ ] Minor (when you add functionality in a backward compatible manner)
- [ ] Patch (when you make backward compatible bug fixes)

At the end of the review, please label the PR with the appropriate label: `r.Major`, `r.Minor`, `r.Patch`.

---

Breaking change release note preparation (if applicable):
- To be included in the release note for any feature that is made obsolete/breaking

> Give a brief explanation note about:
>   - what was the old feature that was made obsolete
>   - any replacement feature (if any), and
>   - how the author should modify his website to migrate from the old feature to the replacement feature (if possible).

This release deprecates support for Vue 2 and introduces full support for Vue 3.

Summary of Breaking Changes:
Vue 2 is no longer supported. All components and internal rendering logic now use Vue 3.3.11. Vue 2-based plugins and custom components must be updated to comply with Vue 3 syntax and API. Deprecated HTML5 syntax support is removed, such as `<font>`, `<big>`, and similar tags that are no longer rendered or styled reliably.

Scoped slots API (scopedSlots) has been replaced by the new slots API as per Vue 3 standards. Internal test and SSR setup has been restructured to reflect modern Vue 3 testing libraries and bundling practices.

What Site Authors Need to Do:
If you are using custom Vue components and directives, please migrate them using the [Vue 3 Migration Guide](https://v3-migration.vuejs.org/). If your site uses deprecated HTML tags, replace them with semantic HTML5 alternatives or appropriate CSS styling. If you have extended MarkBind through plugins that rely on Vue APIs, refer to updated documentation for best practices post-migration.

gerteck commented:
Site is live and running on Vue 3!

Preview at: [https://deploy-preview-2622--markbind-master.netlify.app/](https://deploy-preview-2622--markbind-master.netlify.app/)

Go to the console and type: `Vue.version`. We're running on `Vue version 3.3.11`. 

Ready for review - in the meantime, will try to author a separate PR to aid with PR-review.

EltonGohJH commented:
Hi hi @gerteck,
As mentioned, I think we could deploy a Netlify version of CS2103T for @damithc to test.

Prof @damithc can do that as a first layer of sanity test IMO.

EltonGohJH commented:
Anyways, thanks @gerteck for continuing this feature. Some of the solutions to the problem I had were ingenious, to say the least.

I will try to review and comment on the core SSR flow but likely need someone to help with a in depth check on the other parts like whether all the vue components are functioning.

CC @lhw-1

tlylt commented:
@gerteck great work 🚀 
Just started playing around with the deployed site to check for any defects. One minor thing I noticed when comparing against https://markbind-master.netlify.app/
- The quiz's next icon animates from the top left to bottom right when clicked.

Please see video comparing the deployed preview and current master

https://github.com/user-attachments/assets/78e9d1c8-fff9-46c7-ab97-f82d75cbed55

EltonGohJH commented on packages/core/src/Page/PageVueServerRenderer.ts:
```diff
@@ -3,49 +3,52 @@
-  const outputContent = `
-    var pageVueRenderFn = ${compiledVuePage.render};
-    var pageVueStaticRenderFns = [${compiledVuePage.staticRenderFns}];
+async function compileVuePageCreateAndReturnScript(
+  content: string, pageConfig: PageConfig, pageAsset: PageAssets) {
+  const compilerOptions: CompilerOptions = {
+    runtimeModuleName: 'vue',
+    runtimeGlobalName: 'Vue',
+    mode: 'function',
+  };
+
+  const templateOptions: SFCTemplateCompileOptions = {
+    source: content,
+    filename: pageConfig.sourcePath,
+    id: pageConfig.sourcePath,
+    compilerOptions,
+  };
+  const compiled = compileTemplate(templateOptions);
+  const outputContent = compiled.code;
+
+  const scriptContent = `
+    const renderFn = new Function(${JSON.stringify(outputContent)});
+    var render = renderFn();
```
This part looks good to me. I think is a very good implementation of what we really have.

EltonGohJH commented on packages/vue-components/src/Retriever.vue:
```diff
@@ -92,13 +93,14 @@ export default {
-          const TempComponent = Vue.extend({
+          const TempComponent = {
             template: `<div>\n${result}\n</div>`,
             data() {
               return rootData;
             },
-          });
-          new TempComponent().$mount(this.$el);
+          };
+          const app = createApp(TempComponent);
+          app.mount(this.$el);
```
This part looks good to me.
This is needed because we only render dynamically for this component.

gerteck replied:
I realized I forgot to register the custom components and directives in this new Vue instance, causing bugs 😢

EltonGohJH commented on packages/core/src/Page/PageVueServerRenderer.ts:
```diff
@@ -61,45 +64,60 @@ async function compileVuePageAndCreateScript(
-  const VueAppPage = new Vue({
-    render(createElement: any) {
-      return compiledVuePage.render.apply(this, createElement);
-    },
-    staticRenderFns: compiledVuePage.staticRenderFns,
-    components,
-    directives,
+  // Pass in Vue epxected to be available globally
+  // eslint-disable-next-line @typescript-eslint/no-implied-eval
+  const render = new Function('Vue', renderFn)(Vue);
+  const app = createSSRApp({
+    render,
     ...appFactory(),
   });
+  app.use(plugin);
+  const html = await renderToString(app);
+  const wrappedHtml = `<div id="app">${html}</div>`;
+  return wrappedHtml;
+}
```
LGTM  seems quite straight forward of using SSR to render

EltonGohJH commented on packages/core-web/src/index.js:
```diff
@@ -170,44 +173,32 @@ window.handleSiteNavClick = function (elem, useAnchor = true) {
-    staticRenderFns: pageVueStaticRenderFns,
+  // render exist in global window object
+  // see PageVueServerRenderer.ts
+  const app = createSSRApp({
```
@gerteck 
Same as above

gerteck commented on packages/core-web/src/index.js:
```diff
@@ -170,44 +173,32 @@ window.handleSiteNavClick = function (elem, useAnchor = true) {
+    render,
     ...appFactory(),
     mounted() {
       executeAfterMountedRoutines();
     },
   });
-  /*
-   * For SSR, if we mount onto the wrong element (without data-server-rendered attribute) in our SSR setup,
-   * hydration will fail silently and turn into client-side rendering, which is not what we want.
-   * Thus, we will always force hydration so that we always know when hydration has failed, so that we can
-   * address the hydration issue accordingly.
-   */
-  vm.$mount('#app', true); // second parameter, 'true', enables force hydration
+  app.use(plugin);
+  app.mount('#app');
```
To set some context - from what I understand in differences between `createApp()` and `createSSRApp()` is whether the app runs in `hydration mode`. Reference: [Vue docs](https://vuejs.org/guide/scaling-up/ssr.html#rendering-an-app)
> To make the client-side app interactive, Vue needs to perform the hydration step. During hydration, it creates the same Vue application that was run on the server, matches each component to the DOM nodes it should control, and attaches DOM event listeners.
> To mount an app in hydration mode, we need to use [createSSRApp()](https://vuejs.org/api/application#createssrapp) instead of createApp():

This should be called client side - `index.js` is the entry point for the markbind binary that is bundled in `core-web`. Since it is clientSide - definitely should be using `CreateSSRApp()` - as the client is also served the `*.html` that was pre-rendered, and we want the client to do hydration and check differences betweeen served `html` and DOM structure given by the render function. 

--- 

A related question is whether we should be calling `createApp()` or `createSSRApp()` when we render the `html` in the server. (that is to be sent to client). Referring to this portion in `PageVueServerRenderer`:

https://github.com/gerteck/markbind/blob/7ea3d6d544e7b0d3b69333f2d88279d793d67da5/packages/core/src/Page/PageVueServerRenderer.ts#L88-L103

Theoretically, it should be fine to use `createApp()` on server to render HTML since there's no hydration involved, but I couldn't find any information on this. Following the [serverRenderer (renderToString) npm usage docs](https://www.npmjs.com/package/@vue/server-renderer), it seems that it is supposed to be used with `createSSRApp()` so there is that.

gerteck commented:
> @gerteck great work 🚀 Just started playing around with the deployed site to check for any defects. One minor thing I noticed when comparing against https://markbind-master.netlify.app/
> 
> * The quiz's next icon animates from the top left to bottom right when clicked.


Thank you for your help on this PR! I also noted that [popovers](https://deploy-preview-2622--markbind-master.netlify.app/userguide/components/popups#popovers) also lost their spacing:

![Picture1](https://github.com/user-attachments/assets/d89473dd-5c3f-4c6d-a77c-6d326f6e525e)

It's could be because the vue component libraries were updated or some differences in the configuration syntax, since I made minimal changes enough just to satisfy any errors and warnings.

gerteck commented:
> Hi hi @gerteck, As mentioned, I think we could deploy a Netlify version of CS2103T for @damithc to test.

Thank you @EltonGohJH for making time to look through the PR! Would not have been able to work on this without your previous work on it. Agreed on the suggestion.

gerteck commented:
Currently I forked the CS2103 website and deployed it on my own gh-pages as like a quick draft to view it running on Vue 3 for now, to check for any major regressions first.

Can access it here: [https://gerteck.github.io/cs2103-website/index.html](https://gerteck.github.io/cs2103-website/index.html)
Can always verify the Vue version by running `Vue.version` in the console. Since the site build is running locally I used this MarkBind's vueMigration branch to generate the served pages.

I didn't update the Agolia plugin API Keys and stuff so it will redirect to the original site.

tlylt commented on packages/vue-components/src/SearchbarPageItem.vue:
```diff
@@ -102,7 +102,7 @@ export default {
         white-space: normal;
         vertical-align: top;
-        word-break: break-word;
+        word-break: break-all;
```
why do we need to change this?

gerteck replied:
Somewhere during the migration process I think there was a linting or style error saying that `break-word` was depreciated, so I changed it to fix it then:

I believe it was this error:
https://rocketvalidator.com/html-validation/css-the-value-break-word-is-deprecated
https://stackoverflow.com/questions/47413677/alternative-for-word-break-break-word

However I don't see it anymore if i change it back. Not sure why.

tlylt commented on packages/vue-components/src/__tests__/Breadcrumb.spec.js:
```diff
@@ -4,7 +4,7 @@ import Breadcrumb from '../Breadcrumb.vue';
 const SITENAV_START = `
 <nav id="site-nav">
-<div class="nav-component slim-scroll>
+<div class="nav-component" slim-scroll>
```
```suggestion
<div class="nav-component slim-scroll">
```
i think slim-scroll is a classname

tlylt commented on packages/core/src/Page/PageVueServerRenderer.ts:
```diff
@@ -3,49 +3,52 @@
 import * as logger from '../utils/logger';
 import type { PageConfig, PageAssets } from './PageConfig';
 import type { Page } from '.';
-/* eslint-enable import/no-import-module-exports */
```
let's keep this line? (though the explanation at line 3 is no longer up-to-date

gerteck replied:
okay! I must have accidentally deleted it. Thanks!!

tlylt reviewed (commented):
Some more issues I noticed on the rendered site:
- The tree plugin is no longer working:
  - ![Screenshot 2025-03-13 at 9 19 05 PM](https://github.com/user-attachments/assets/60b7bbd3-6e3b-4996-a39d-277bae2c01b9)
- Modal opening animation is gone (before: a smooth transition to the opened modal, now: immediate opening of modal window)

damithc commented:
One issue I noticed: 2nd level drop-down menus do not disappear as they should

![image](https://github.com/user-attachments/assets/b3b0aa25-f510-4fde-bfd8-35cdbb5771e6)

tlylt commented:
Tried serving the markbind docs locally with this branch and got this warning:
```
 v5.5.3
info: Website generation started at 6:45:41 AM
info: Building assets...
info: Assets built
info: Generating pages...
[------------------------------------------------------------] 0 / 60 pages builtunsafe attribute name: tim"
[@vue/server-renderer] Skipped rendering unsafe attribute name: tim"
unsafe attribute name: xu"
[@vue/server-renderer] Skipped rendering unsafe attribute name: xu"
[============================================================] 60 / 60 pages built
```
due to 
```
      <td align="center" valign="top" width="14.28%"><a href="http://tim-siu.github.io"><img src="https://avatars.githubusercontent.com/u/61866948?v=4?s=100" width="100px;" alt="Shuyao "Tim" Xu"/><br /><sub><b>Shuyao "Tim" Xu</b></sub></a><br /><a href="https://github.com/MarkBind/markbind/commits?author=Tim-Siu" title="Code">💻</a></td>
```

this is generated code so probably out of our control, but am wondering if we have a way to disable check for vue server-renderer just for this part..

gerteck commented:
> One issue I noticed: 2nd level drop-down menus do not disappear as they should

Identified that the issue was due to a [breaking change in Vue 3](https://v3-migration.vuejs.org/breaking-changes/children)

I have made a non-trivial change to the logic to hide other submenu instances at the same level, shifting from the use of `$children` to maintaining a list of submenus at each level using `provide/inject` to resolve this issue.

gerteck commented:
> Some more issues I noticed on the rendered site:
> 
> * The tree diagram functionality is no longer working:
>   
> * Modal opening animation is gone (before: a smooth transition to the opened modal, now: immediate opening of modal window)

I have looked into the tree diagram functionality: The root cause is due to changing of default handling of whitespaces from Vue 2 to Vue 3, specifically: [https://vuejs.org/api/application.html#app-config-compileroptions-whitespace](https://vuejs.org/api/application.html#app-config-compileroptions-whitespace)
* default in Vue 2 is to preserve, changed to condense in Vue 3.

A few ways we can approach this
* set default option to `preserve` for the Vue app in render function generation, rendering and mounting
* transform tree element to `<pre class="tree"/>`, as `<pre/>` use case is to preserve all whitespaces
* do both.

Will investigate modal opening animation soon alongside that of quiz button animation

tlylt commented on packages/vue-components/src/questions/Question.vue:
```diff
@@ -80,7 +80,7 @@
           </button>
           <button
-            v-if="qState.state !== 0 && questions"
+            v-show="qState.state !== 0 && questions"
```
Just found out why the quiz is not working well...probably due to transition groups that no longer work as expected due to this change such that the elements appearance sequence is different.

You mentioned that this is to fix a hydration issue and when I tested locally by changing v-show back to v-if, it indeed triggered hydration warnings in the console. I looked into this further as we don't fetch this question from a server, so I am curious why it is not able to have a consistent state between server and client.

After a few tries, the following changes seems to be the fix:
- in QuizQuestionMixin.js
```
  inject: {
    questions: {
      default: [], // or () => [], need to check the type again
    },
```

gerteck replied:
Currently investigating this issue, and I can't seem to figure out the hydration error exists yet as well.

Because of how `Quiz` and `Question` both check if `questions` is defined or undefined to check if the question is in a quiz, the change to default `questions` to `[]` breaks some logic in the question component.

I will do some more investigation to either try to uncouple it or provide some workaround.

tlylt replied:
> Because of how Quiz and Question both check if questions is defined or undefined to check if the question is in a quiz, the change to default questions to [] breaks some logic in the question component.

i see, how about changing those logic to check if the question array is of length 0? (i didn't check, just thinking out loud)

gerteck replied:
I tried that too but unfortunately I realized its quite coupled (q a bit of logic depends on question being defined or undefined)

https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/packages/vue-components/src/questions/QuizQuestionMixin.js#L44-L49

(^ here, the question immediately gets pushed into the array if the array is defined so the length would not be 0)

Also I don't think it hits at the core of the issue of the hydration error, so refactoring might not fix the hydration error - though I think refactoring would definitely not be bad, but maybe expand the PR scope too much

tlylt replied:
> Because of how Quiz and Question both check if questions is defined or undefined to check if the question is in a quiz, the change to default questions to [] breaks some logic in the question component, so it can't be used.

could you share what logic was broken?

gerteck replied:
To set the context, In the current logic, the only way for the `questions` array to be defined is if the quiz was a parent component and initialized the questions array. So the quiz being undefined is used in the logic to check whether it was in a quiz or not.

![image](https://github.com/user-attachments/assets/306cb047-15f1-49ca-ab58-020d6f72bdc4)

To elaborate specifically why, the quiz would `provide` a defined array [], so if our `injected` question was defined, it must mean that the question is in a parent quiz component.

https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/packages/vue-components/src/questions/Quiz.vue#L116-L130

This affects the logic in the snippet in previous comment. If we define the default value of the questions array to be a defined, but empty array. The question is pushed into the `questions` array (through [provide/inject](https://vuejs.org/guide/components/provide-inject) by Quiz), because it thinks it is a parent Quiz component.  At the same time, it hides the question** as well. (To be shown by the quiz component as per quiz logic later)

https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/packages/vue-components/src/questions/QuizQuestionMixin.js#L44-L49

```js
 created() { 
   if (this.questions) {               // thinks it is in a quiz
     this.active = false;              // hides the question
     this.questions.push(this);   // adds this question in the array
   } 
 }, 
```

This raises some issues because some methods in `question.vue` uses the property of whether `questions` is defined to determine if it is in the quiz. I think it could be refactored to make it more clear and have proper methods as this is logical dependency is not so clear at first (or second) glance.

https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/packages/vue-components/src/questions/Question.vue#L150-L163

However, a more relevant issue is that I don't think it resolves the hydration issue. I noted that the reason why the hydration error doesn't show up is because if we set `questions:` `default: []`, as per the logic in `created()` above is because it hides the question. `active=false`. Specifically this causes the whole question to not render (`v-if=active`):

https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/packages/vue-components/src/questions/Question.vue#L1-L3

 But actually if we set it back to `active=true` actually the hydration error still persists even if `questions` array is defined..

tlylt commented:
> * default in Vue 2 is to preserve, changed to condense in Vue 3.

Great finding, yup i just searched the web and it seems like a known issue https://github.com/vuejs/core/issues/2431

I think if there are more stuff broken because of this, we can try adjusting the compiler option, if not we can probably fix the tree functionality, which i suppose should be doable.

gerteck commented:
> > * default in Vue 2 is to preserve, changed to condense in Vue 3.
> 
> Great finding, yup i just searched the web and it seems like a known issue [vuejs/core#2431](https://github.com/vuejs/core/issues/2431)
> 
> I think if there are more stuff broken because of this, we can try adjusting the compiler option, if not we can probably fix the tree functionality, which i suppose should be doable.

I just tested adjusting the compiler option, it works with just a simple tweak:
![image](https://github.com/user-attachments/assets/b6a3ee84-9a30-4b26-a5e7-1c3599331849)

Result:


Though the documentation for the compilerOptions is non-existent, so would need to refer to the codebase for the options: https://github.com/vuejs/core/blob/main/packages/compiler-core/src/options.ts

tlylt commented:
> I just tested adjusting the compiler option, it works with just a simple tweak:

I'm ok with this change since we preserve existing behavior, we can create a separate issue to track this to explore changing it to condense

tlylt commented:
> transform tree element to `<pre class="tree"/>, as <pre/>` use case is to preserve all whitespaces

just a note, i now remembered that i probably intentionally not use `pre` because we allow markdown syntax, so there could be elements within a parent div, but not ideal if the parent is `pre`

gerteck commented:
> The quiz's next icon animates from the top left to bottom right when clicked.
> Just found out why the quiz is not working well...probably due to transition groups that no longer work as expected due to this change such that the elements appearance sequence is different.
> You mentioned that this is to fix a hydration issue and when I tested locally by changing v-show back to v-if, it indeed triggered hydration warnings in the console. 

I think regarding the Question hydration error - I can revert the changes and let the hydration warning lapse for now.

An alternative could be to find a workaround to avoid this, though not sure how feasible nor if worth it.

Upon investigation, it could be a internal vue issue during rendering - See vue/6715 (https://github.com/vuejs/core/issues/ 6715). This seems to be patched in a later vue version (after v3.3.11). 

So we can choose to update Vue to latest version. (Currently using v3.3.11). However, I faced issues doing so, seemingly related to typescript:
![image](https://github.com/user-attachments/assets/2016628c-4c17-47ae-86a7-c3ff4d237dce)

It could be due to outdated typescript version, some typing library issue or something else. I previously tried to resolve this issue but I couldn't make enough headway as the scope easily expanded beyond what was comprehensible in a single PR. I also raised some issues which don't seem very trivial to debug. 

Hence -  I propose I revert the changes to questions, and let the hydration warning lapse for now. (The hydration warning is quite minor, maybe only causes slight rendering inefficiency in this case as it does not cause it to rerender, and everything still works.)
Then, we raise an issue to update the vue version as well as typescript errors and set it to higher priority.

gerteck commented:
> Tried serving the markbind docs locally with this branch and got this warning (Skipped rendering unsafe attribute name)
> this is generated code so probably out of our control, but am wondering if we have a way to disable check for vue server-renderer just for this part..

I couldn't find a way to disable the warnings for server-renderer meaningfully. However, the warning does point out a legitimate issue though — the HTML generated by the All Contributors bot has unescaped quotes in the alt attribute:
```
<img src="URL" width="100px;" alt="Shuyao "Tim" Xu"/>
```
https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/docs/about.md?plain=1#L133
The issue is occuring because the inner quotes around are not escaped (i.e., should be &quot;). Causing the html to be recognised as `alt = "Shuyao "` and `Tim` and `Xu"`. The error in the code isn't generated by vue renderer, but rather the all-contributors bot. 

Since this is an edge case, maybe instead we can open a small PR to manually escape the quotes in about.md (i.e., alt="Shuyao &quot;Tim&quot; Xu"). Also, maybe file a PR or issue upstream with the All Contributors project to raise concern on handling alt attributes with quotes.

Tim-Siu commented:
> > Tried serving the markbind docs locally with this branch and got this warning (Skipped rendering unsafe attribute name)
> > this is generated code so probably out of our control, but am wondering if we have a way to disable check for vue server-renderer just for this part..
> 
> I couldn't find a way to disable the warnings for server-renderer meaningfully. However, the warning does point out a legitimate issue though — the HTML generated by the All Contributors bot has unescaped quotes in the alt attribute:
> 
> ```
> <img src="URL" width="100px;" alt="Shuyao "Tim" Xu"/>
> ```
> 
> The issue is occuring because the inner quotes around are not escaped (i.e., should be & quot;). Causing the html to be recognised as `alt = "Shuyao "` and `Tim` and `Xu"`. The error in the code isn't generated by vue renderer, but rather the all-contributors bot.
> 
> https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/docs/about.md?plain=1#L133
> 
> Since this is an edge case, maybe instead we can open a small PR to manually escape the quotes in about.md (i.e., alt="Shuyao "Tim" Xu"). Also, maybe file a PR or issue upstream with the All Contributors project to raise concern on handling alt attributes with quotes.

I will file a PR to fix this 😭

Tim-Siu commented:
> > Tried serving the markbind docs locally with this branch and got this warning (Skipped rendering unsafe attribute name)
> > this is generated code so probably out of our control, but am wondering if we have a way to disable check for vue server-renderer just for this part..
> 
> I couldn't find a way to disable the warnings for server-renderer meaningfully. However, the warning does point out a legitimate issue though — the HTML generated by the All Contributors bot has unescaped quotes in the alt attribute:
> 
> ```
> <img src="URL" width="100px;" alt="Shuyao "Tim" Xu"/>
> ```
> 
> The issue is occuring because the inner quotes around are not escaped (i.e., should be `&quot;`). Causing the html to be recognised as `alt = "Shuyao "` and `Tim` and `Xu"`. The error in the code isn't generated by vue renderer, but rather the all-contributors bot.
> 
> https://github.com/MarkBind/markbind/blob/ea7cf57e067f0880d470265fad2fae8a0d3eaa24/docs/about.md?plain=1#L133
> 
> Since this is an edge case, maybe instead we can open a small PR to manually escape the quotes in about.md (i.e., alt="Shuyao `&quot;`Tim`&quot;` Xu"). Also, maybe file a PR or issue upstream with the All Contributors project to raise concern on handling alt attributes with quotes.

Thanks for the investigation. Indeed this seems to be All Contributors bot not escaping the quote. I am not sure if we can manually modify it though. The header mentions `<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->`

gerteck commented:
> * Modal opening animation is gone (before: a smooth transition to the opened modal, now: immediate opening of modal window)

Thanks for spotting this! Was due to some naming changes transition in Vue and the package.
* https://v3.vue-final-modal.org/guide/properties/#transition
* https://v3-migration.vuejs.org/breaking-changes/transition

Have pushed changes

tlylt commented:
> I think regarding the Question hydration error - I can revert the changes and let the hydration warning lapse for now.
> An alternative could be to find a workaround to avoid this, though not sure how feasible nor if worth it.

What about removing the animation altogether and create a separate issue for re-implementing that instead?

 > Thanks for the investigation. Indeed this seems to be All Contributors bot not escaping the quote. I am not sure if we can manually modify it though. The header mentions <!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section --> Maybe we can have a separate issue addressing this if deemed important.

Yes given the state of the all contributor project it doesn't seem likely that they will merge a fix from us. Manual fix might be overridden when the page is auto-updated. Might need to test it out because I wonder if they really update the username on every request...If they don't, then we can manually update this html. If they do...unless we request you to change your username...we can live with this warning for now.

gerteck commented:
Updated Links for convenience:

* https://deploy-preview-2622--markbind-master.netlify.app/
* https://gerteck.github.io/cs2103-website/se-book-adapted/index.html
  * (https://github.com/gerteck/cs2103-website)

Aggregated Changes since previous:
* Modals animation updated
* Nested Dropdowns hide other submenus
* Trees work
* Spaces between popovers updated

tlylt commented:
A few more issues on mobile
- the navbar content is not properly handled
- weird white space when opening the site nav or page nav

![image](https://github.com/user-attachments/assets/631a3a48-d818-4e56-bbf6-44ac9b012488)
![image](https://github.com/user-attachments/assets/ab80baa1-6476-4370-83a4-1a73e6193f9e)
![image](https://github.com/user-attachments/assets/38982176-d86e-4d51-bbf4-d802a24e36a2)

tlylt commented:
One more issue on the 2103T website:
- Left is vue3: https://gerteck.github.io/cs2103-website/schedule/week9/topics.html
- Some images are not loaded
- The panel outer layout is somehow missing
![image](https://github.com/user-attachments/assets/7ca5683c-88c6-4491-b951-402e5388724d)

gerteck commented:
> One more issue on the 2103T website:
> 
> * Left is vue3: https://gerteck.github.io/cs2103-website/schedule/week9/topics.html
> * Some images are not loaded
> * The panel outer layout is somehow missing
> * (there is a hydration warning in the console, not sure if it's related)

Noted on the images and panels not loading/missing. I realized this was because I did not register MarkBind's custom Vue components in new Vue instances that were mounted when we dynamically load content, pertinently in `<includes>`. This caused it to not recognise custom vue components such as `<pic>`, `<panel>` in rendering, causing these issues.

To resolve this, I have globally exposed the MarkBindVuePlugin on the window object such that we can grab it and register it in `Retriever.vue` when we initialize a new Vue instance to dynamically load content, such as when we expand panels, or load through `<includes>`.

Have yet to investigate the mobile accessibility issues, will get to it when I can

gerteck commented:
I am investigating the changes in mobile accessibility styling - it looks to be caused by `markbind.min.css` not loading in, either not being bundled or not being accessed properly. Trying to find out what has caused this regression and how to revert it.

gerteck commented:
@damithc I've updated the site with the latest version of the branch: https://gerteck.github.io/cs2103-website/index.html
* I will update the site as issues are fixed https://github.com/gerteck/cs2103-website/deployments

tlylt commented:
@gerteck great work so far!

Could you also include some details of how we integrate Vue in our Dev Guide (under Design, perhaps a new page)? Need not explain everything, but keep at high level and assume that reader understands Vue and MarkBind. I'm looking for how we import/setup the vue app/create render function/some information about CSR SSR, so that this can help future Vue migration work and point out important places to take note of (such as "we initialize a new Vue instance to dynamically load content")
- @EltonGohJH can probably help with reviewing the docs 🙏 

I won't be available until April so I'm going to give my blessing now and support for this PR to be merged once we resolved all critical issues 🚀 🚀

damithc commented:
> I won't be available until April so I'm going to give my blessing now and support for this PR to be merged once we resolved all critical issues 🚀 🚀

Thanks for your inputs so far, @tlylt 💯

gerteck commented:
I have added a new page detailing how MarkBind integrates Vue (SSR, CSR, Hydration, Execution Flow).

* https://deploy-preview-2622--markbind-master.netlify.app/devguide/design/vueintegration

I think it would be useful for reviewers as well, so commenting it here to highlight it.

gerteck commented:
I have investigated and identified the cause of regression for site-nav and page-nav whitespace. 

This was due to a change in `portal-vue` rendering mechanism, specifically [portal-target no longer rendering a root element](https://portal-vue.linusb.org/guide/migration.html#portaltarget-component)
* A high-level explanation - Due to this expectation, the style attribute accounting for the `--sticky-header-height` was applied twice, causing this whitespace.
* Fix: applied a parent div to `portal-target` to 'absorb' this styling to maintain the expectation as per `vue-portal` v2. We can consider refactoring this in the future.  However, note that the change to refactor is not too trivial and requires some significant understanding of current styling in `main.css`, with regards to `#page-nav` and `#site-nav`, and how it interacts with `Overlay`, `OverlaySource`, and `portal-vue`.

damithc commented:
> I have added a new page detailing how MarkBind integrates Vue (SSR, CSR, Hydration, Execution Flow).
> 
> * https://deploy-preview-2622--markbind-master.netlify.app/devguide/design/vueintegration
> 
> I think it would be useful for reviewers as well, so commenting it here to highlight it.

@gerteck Thanks for adding to dev docs. Better dev docs will help future devs, for sure.

Given below are the two versions of the textbook PDF files. This is not exactly the result of 'save as PDF' but a result of a Python script that I used to generate the PDF and add a ToC and an index.

[textbook-vue2.pdf](https://github.com/user-attachments/files/19401955/textbook-vue2.pdf)
[textbook-vue3.pdf](https://github.com/user-attachments/files/19401956/textbook-vue3.pdf)

By the way, your staged CS2103 site says `2020 Aug-Dec`. Perhaps you are using the `master` branch rather than the `cs2103` branch of the website repo? But it doesn't matter much as the differences are minor.
![image](https://github.com/user-attachments/assets/6e732513-ba12-43de-b16c-5abd533b75af)

gerteck commented:
I initially updated the directive's `bind()` hook to `beforeMount()` based on the [official Vue migration guide](https://v3-migration.vuejs.org/breaking-changes/custom-directives.html). However, this change caused the directive to stop working correctly — since the DOM mutations need to happen after the element is mounted.

I've now fixed this by updating the hook from beforeMount() to mounted(), ensuring all DOM manipulations and event bindings happen after the element is fully available in the DOM.

The `closeable` directive works now: https://deploy-preview-2622--markbind-master.netlify.app/userguide/reusingcontents#allowing-users-to-remove-some-contents

gerteck commented:
---
Links for convenience:

* Vue 2 MarkBind Docs: https://markbind-master.netlify.app/
* Vue 2 CS2103 Site https://nus-cs2103-ay2425s2.github.io/website/index.html

* Vue 3 MarkBind Docs:  https://deploy-preview-2622--markbind-master.netlify.app/
* Vue 3 CS2103 Site: https://gerteck.github.io/cs2103-website/se-book-adapted/index.html
  * (https://github.com/gerteck/cs2103-website)

---
Edit: I am including this for easier reference in latest comment

**Changelog** (Beyond initial PR changes):
* Built the CS2103 website on gh-pages (serves updated version)
* Update modal opening animation (update breaking syntax)
* Update dropdown functionality (hide submenus - update breaking syntax with new logic)
* Fix tree diagram functionality (whitespace issue - vue breaking change)
* Fix spacing between elements (whitespace issue - vue breaking change)
* Investigate render warning (invalid HTML in docs)
* Fix to images not loaded, panel somehow missing (register plugins in new Vue instance, Retriever.vue)
* Updated fix to navbar content not properly handled (update breaking vue CSS scoped syntax)
* Update Question hydration error - (disable animations for this PR for now)
* Add docs for how MarkBind uses Vue & Update outdated docs on vue 
* Fix whitespace issue in site-nav / page-nav (update portal-vue breaking syntax)
* Closeable Directive fixed (update hook)

**To Do**:
* Investigate hydration errors in served CS2103 site.

**Issues outside scope of this PR**:
* Re-enable animations for Question
* Refactor styling for site-nav, page-nav, overlaysource
* Refactor closable directive (issue has been made)

gerteck commented:
I have found some panels not loading:

![image](https://github.com/user-attachments/assets/51e145f3-f030-40a2-a9e6-02cf17c0f44c)

However, looking into the cs2103 website code, I am not too sure how the panel is including the src:

![Picture1](https://github.com/user-attachments/assets/1b42dc97-db22-4ffa-bc2a-d1828f046d3c)

Could I be setting up the website wrongly?

damithc commented:
> I don't think i understand how `unit-inElsewhere` works, so I can't really debug it

@gerteck This is a case of using a feature called `boilerplate` (that feature saves us from having to duplicate a boilerplate file in multiple folders). The `unit-inElsewhere-asFlat.md` is actually in the `_markbind\boilerplates` folder but MarkBind is supposed to act as if the file is in the the folder specified in front of it (in this case, `../../../uml/sequenceDiagrams/parallelPaths`).

Incogdino commented:
![image](https://github.com/user-attachments/assets/d82c8057-20f0-4c7a-a0d5-a23338106c34)

The images appear to be loading at max width on the vue 3 site as compared to the 2103 site

lhw-1 commented:
Ah, same observation as @Incogdino. Left is on vue 2, right is on vue 3.

![image](https://github.com/user-attachments/assets/16a8adbf-edc6-442f-813e-b45612e560c9)

lhw-1 commented:
I've taken a closer look at both of the CS2103T site deployments, and apart from the image size issue and the previously mentioned site-nav and page-nav whitespaces (which we can tackle outside of this PR), there does not seem to be major issues.

As discussed with @damithc, we're pushing for this to be merged in within the next 2 days, after which we will release MarkBind v6.0.0. @MarkBind/active-devs let's proceed with the final round of checks and testing, and @gerteck do let us know your plans for this PR if there are any remaining.

gerteck commented:
> I've taken a closer look at both of the CS2103T site deployments, and apart from the image size issue and the previously mentioned site-nav and page-nav whitespaces (which we can tackle outside of this PR, as per [comment](https://github.com/MarkBind/markbind/pull/2622#issuecomment-2744983002)), there does not seem to be major issues.
> 
> As discussed with @damithc, we're pushing for this to be merged in within the next 2 days, after which we will release MarkBind v6.0.0. @MarkBind/active-devs let's proceed with the final round of checks and testing, and @gerteck do let us know your plans for this PR if there are any remaining.

Thank you for your help!

I am looking into the image size issue, I have isolated and identified the regression, and can push a fix in the next hour or so.

Otherwise, my plan would be that any other changes. if not major or functionality breaking can be left to be fixed after the PR has been merged.

gerteck commented:
![pic](https://github.com/user-attachments/assets/bd62ae38-6174-4716-b323-4d68237aebb3)

The pictures did not resize and defaulted to max size was because the method to set height and width was not firing, which were set to fire on `load` event. It did not fire on Vue 3 hydration, but did so for Vue 2.

The reason the load event fired in Vue 2 but not in Vue 3 is likely due to a difference in hydration behavior between Vue 2 and Vue 3 during SSR (Server-Side Rendering). In short, it seems like:
* Vue 2: load event fired after hydration of DOM.
* Vue 3: likely the onload event is fired before Vue runs and hydrates the DOM.

This seems to be a somewhat known (?) issue through search.

It could be that that when the component is hydrated in Vue 3, the image is already in the DOM and is marked as loaded and not fire `load` event for performance reasons. Or that Vue 2 retriggers the load event etc. I don’t know the exact reason.

[Load Event](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event)

What I’ve done is to add a `onMount` routine to run the height and width calculation and setting if the image is already loaded, to be sure that it runs in any case.

It also seems to be used in `Annotate`, `Thumbnail`, let me check it and push changes if needed as well

IanCheah commented:
Hi @gerteck, I have noticed a little difference under Admin Info/Ip/Week2.
Vue 3: [https://gerteck.github.io/cs2103-website/admin/ip-w2.html](https://gerteck.github.io/cs2103-website/admin/ip-w2.html)
Vue 2: [https://nus-cs2103-ay2425s2.github.io/website/admin/ip-w2.html](https://nus-cs2103-ay2425s2.github.io/website/admin/ip-w2.html)
![image](https://github.com/user-attachments/assets/e88027f2-c684-497d-bacd-ea0988f45f44)
On the left is Vue3, right is Vue2. There seem to be a little issue with the rendering of the mermaid diagram.
![image](https://github.com/user-attachments/assets/51bad86f-fb61-4983-ab43-2f9da0c8ad90)

gerteck commented:
> Hi @gerteck, I have noticed a little differences under Admin Info/Ip/Week2.

Thanks @IanCheah for your help! I will have to investigate the tabs issue. 

Regarding the mermaid rendering, it seems related to CORS and seems to be quite flaky, it could have been a regression from my previous PR shifting mermaid to load on demand. I can only replicate the issue sometimes. It seems that some investigate on that is due as well, though I suppose if we support offline loading and bundle it together it should totally resolve the issue, which is related to #2548

Should I work on it now, or should we postpone it till after the migration, and set up a separate issue to tackle this?

IanCheah commented:
> I think we should postpone this issue till after the migration, and set up a separate issue to tackle this? Otherwise, any more commits will continue to balloon this PR. What do you think?

Yep I agree! Adding more commits may balloon the pr and even potentially introduce unforseen errors. We can set up a separate issue once the pr is merged.

lhw-1 reviewed (approved):
Thanks for all your work, @gerteck! 

This has no doubt been a massive undertaking on your part, and your efforts in documenting all the steps are greatly appreciated - I'm sure future devs will benefit from your efforts. Thanks to @EltonGohJH who started this effort 2 years ago, as well as to all the devs who helped to review and test this PR as well (@tlylt, @Tim-Siu, @Incogdino, @IanCheah, and of course @damithc)!

LGTM, and if there are no other updates, we can proceed with the planned release tonight :+1:

damithc commented:
> I think we should postpone this issue till after the migration, and set up a separate issue to tackle this? Otherwise, any more commits will continue to balloon this PR. What do you think?

Yup, that should be fine. But it needs to be fixed before releasing V6, as it breaks current usages.

damithc commented:
> This has no doubt been a massive undertaking on your part, and your efforts in documenting all the steps are greatly appreciated - I'm sure future devs will benefit from your efforts. Thanks to @EltonGohJH who started this effort 2 years ago, as well as to all the devs who helped to review and test this PR as well (@tlylt, @Tim-Siu, @Incogdino, @IanCheah,

I second that 💯 👍
